Credit error

stijnvanhoey opened the pull request:
This PR tackles #80 
 by capturing the return XML message returned when the credit limit is reached and providing an extensive error message explaining how to request a token.

When the credit limit is exceeded, the user gets the following error:

```
Error: API did not return json - Credit limit exceeded for type IP and key 213.49.33.200 When you
require more extended data requests, please request a download token from the waterinfo.be site
administrators via the e-mail address hydrometrie@waterinfo.be with a statement of which data and
how frequently you would like to download data.
```

@rogiersbart thanks for reporting. Would this information be sufficient?

@peterdesmet or @damianooldoni I do not directly know how to properly unit test this adjustment?  Any suggestions?

rogiersbart commented:
That would definitely be a huge improvement! Not sure if this is commonly done, or if you would be willing to, but I would also appreciate including the link https://docs.ropensci.org/wateRinfo/#note-on-restrictions-of-the-downloads or pointing to `?wateRinfo::token`, so people see that there is some documentation on this inside the package. Furthermore, stating this token can be used in the data request function calls might be useful as well. There is an example in the link, not really in the token help page (adding that example would be useful too I think, and then the explanation should not be part of the error message).

stijnvanhoey commented:
ok, good suggestions to take into account. As links do tend to change (and I will forget to update them inside the code), I'm not really into adding the link to the webpage. So, I added the `?wateRinfo::token` suggestion and added the data request example to the documentation of it. Thanks!

damianooldoni commented:
Nice work! About testing, I would use `expect_error()` with the right returned message. To do so, maybe you could use a kind of dummy key created by administrator simulating the credit limit exceeded. I don't think it is something you can do alone just on the package side.

In some cases, a useful turn-around could be using [`vcr` R package](https://docs.ropensci.org/vcr/index.html which helps you stub and replay HTTP requests. It's something rgbif uses extensively for testing more complex and heavy queries. In this way you just need to get the error only the first time: the output is saved and then used from second time. Disadvantage: you need to delete the recorded output if something changes at admin level, but it doesn't happen so often.

rogiersbart commented:
> ok, good suggestions to take into account. As links do tend to change (and I will forget to update them inside the code), I'm not really into adding the link to the webpage. So, I added the `?wateRinfo::token` suggestion and added the data request example to the documentation of it. Thanks!

Yes, makes sense. Thanks to you for fixing this!

stijnvanhoey commented:
@damianooldoni thanks for the advice. I'm merging this. We do not have the ability to change the setup on server side (@WillemMaetens if there is any possibility like that, would be welcome to adjust).

WillemMaetens commented:
Did not check this often enough apparently, I can provide a token with a low credit allowance for simulating this if necessary.

stijnvanhoey commented:
@WillemMaetens I do think this could indeed help for the CI checks. I'll save the token as a CI github action secret and make it as such available for the tests, that would probably work. Thanks!

WillemMaetens commented:
There are two tokens that can be used:

- test_100 has 100 credits/day and should be OK for testing calls and can be exhausted by doing multiple calls for testing purposes if necessary: MzJkY2VlY2UtODI2Yy00Yjk4LTljMmQtYjE2OTc4ZjBjYTZhOjRhZGE4NzFhLTk1MjgtNGI0ZC1iZmQ1LWI1NzBjZThmNGQyZA==

- test_0 hast 0 credits/day and should result in an error on the second call (the first call will work as the first will exceed the allowance):  YWNjNmFjOWYtZmVhNS00MmM5LWE4MWItZjZiMTk4NWE5ZGFjOmUzNWFjMjllLWIzMDQtNDY3My1iMGNjLWUyMTg0MmRjMDk2Mg==

However, it does take some time for the system to recognize that a token is exhausted and even the token with 0 credits may work for some period of time after a first call has been made before errors start being returned.
